File: packages/sdk/examples/learning/submit.py
"""python3 submit.py SOURCE_ID example.json (Python standard library only)."""
import json
import os
import sys
import urllib.error
import urllib.parse
import urllib.request


class NoRedirect(urllib.request.HTTPRedirectHandler):
    def redirect_request(self, req, fp, code, msg, headers, newurl):
        return None


def main():
    if len(sys.argv) != 3:
        raise ValueError("Usage: submit.py SOURCE_ID example.json")
    source_id, filename = sys.argv[1:]
    endpoint = os.environ["OPENPOND_API_URL"].rstrip("/")
    parsed = urllib.parse.urlparse(endpoint)
    if parsed.scheme not in ("http", "https") or not parsed.hostname or parsed.username or parsed.password or parsed.query or parsed.fragment:
        raise ValueError("Use an HTTP(S) API URL without credentials, query or fragment.")
    scope = os.environ["OPENPOND_LEARNING_SCOPE"]
    token = os.environ["OPENPOND_API_KEY"]
    opener = urllib.request.build_opener(NoRedirect())

    def post(route, payload):
        body = json.dumps(payload, allow_nan=False).encode("utf-8")
        if len(body) > 16_777_216:
            raise ValueError("Request exceeds 16 MiB.")
        request = urllib.request.Request(endpoint + "/v1/learning/" + route, data=body, headers={"Authorization": "Bearer " + token, "Content-Type": "application/json", "Accept": "application/json", "X-OpenPond-Team-Id": scope}, method="POST")
        try:
            with opener.open(request, timeout=60) as response:
                result = response.read(16_777_217)
                if len(result) > 16_777_216:
                    raise ValueError("Response exceeds 16 MiB.")
                return json.loads(result)
        except urllib.error.HTTPError as error:
            # Do not print request headers or echoed evidence from error bodies.
            raise RuntimeError("Learning API rejected the request (HTTP %s)." % error.code) from None

    source = post("read", {"action": "get", "scope": scope, "kind": "source", "id": source_id})
    with open(filename, encoding="utf-8") as handle:
        example = json.load(handle)
    example.update(schemaVersion="openpond.taskExample.v1", sourceId=source_id, taskDefinition=source["taskDefinition"])
    operation_id = example["idempotencyKey"]
    result = post("commands", {"scope": scope, "command": {"action": "submit_example", "operationId": operation_id, "example": example}})
    if result["operationId"] != operation_id:
        raise RuntimeError("Operation identity mismatch.")
    evidence = result["resources"][0]
    if evidence["schemaVersion"] != "openpond.taskEvidence.v1":
        raise RuntimeError("Unexpected resource type.")
    print(json.dumps({"operationId": operation_id, "evidenceId": evidence["id"], "revision": evidence["revision"], "contentHash": evidence["contentHash"]}))


if __name__ == "__main__":
    main()

File: packages/sdk/examples/learning/submit.ts
import { readFile } from "node:fs/promises";
import { OpenPondLearningClient, TaskExampleSubmissionSchema, TaskEvidenceSchema } from "openpond-sdk/learning";

// node --experimental-strip-types submit.ts SOURCE_ID example.json
// Set OPENPOND_API_KEY, OPENPOND_API_URL and OPENPOND_LEARNING_SCOPE in your shell.
function required(name: string): string {
  const value = process.env[name];
  if (!value) throw new Error(`Set ${name} before running this example.`);
  return value;
}
const [sourceId, file] = process.argv.slice(2);
if (!sourceId || !file) throw new Error("Usage: submit.ts SOURCE_ID example.json");
const client = new OpenPondLearningClient({ apiKey: required("OPENPOND_API_KEY"), baseUrl: required("OPENPOND_API_URL"), scope: required("OPENPOND_LEARNING_SCOPE") });
const source = await client.sourceConfiguration(sourceId);
const record: unknown = JSON.parse(await readFile(file, "utf8"));
if (!record || typeof record !== "object" || Array.isArray(record)) throw new Error("Expected one JSON object.");
const example = TaskExampleSubmissionSchema.parse({ ...record, schemaVersion: "openpond.taskExample.v1", sourceId, taskDefinition: source.taskDefinition });
const receipt = await client.submitExample(example);
const evidence = TaskEvidenceSchema.parse(receipt.resources[0]);
console.log(JSON.stringify({ operationId: receipt.operationId, evidenceId: evidence.id, revision: evidence.revision, contentHash: evidence.contentHash }));
